prog/weaveproxy/main.go

```diff
@@ -85,3 +87,44 @@ func main() {
 
 	p.ListenAndServe()
 }
+
+func loadDefaultListenAddr(c *proxy.Config) {
+	if len(c.ListenAddrs) > 0 {
+		return
+	}
+
+	clientTLSEnabled := (os.Getenv("DOCKER_CLIENT_TLS_VERIFY") != "")
+
+	clientArgs := strings.Fields(os.Getenv("DOCKER_CLIENT_ARGS"))
+	filteredArgs := []string{}
+	for i := 0; i < len(clientArgs); i++ {
+		arg := strings.SplitN(clientArgs[i], "=", 2)
+		switch arg[0] {
+		case "-tls", "--tls", "-tlsverify", "--tlsverify":
+			clientTLSEnabled = true
+		case "-H", "--host":
+			filteredArgs = append(filteredArgs, clientArgs[i])
+			if !strings.Contains(clientArgs[i], "=") {
+				if i++; i < len(clientArgs) {
+					filteredArgs = append(filteredArgs, clientArgs[i])
+				}
+			}
+		}
+	}
+
+	if clientTLSEnabled && !(c.TLSConfig.Enabled || c.TLSConfig.Verify) {
+		log.Fatalln("Cannot autoconfigure proxy listen address when launching via TLS")
+	}
+
+	host := os.Getenv("DOCKER_CLIENT_HOST")
+	f := &mflag.FlagSet{}
+	f.StringVar(&host, []string{"H", "-host"}, host, "")
+	if err := f.Parse(filteredArgs); err != nil {
+		log.Fatalf("Error configuring proxy: %s", err)
+	}
+
+	c.ListenAddrs = []string{"tcp://0.0.0.0:12375"}
+	if host == "" || strings.HasPrefix(host, "unix://") {
+		c.ListenAddrs = []string{"unix:///var/run/weave.sock"}
+	}
+}
```

Comment:
Most of the code in this whole function should be moved to the script. Passing configuration params to the proxy via env vars is gross. Also, we can do much better error handling in the script, i.e. the user won't just be told that proxy died.